pkg/frontend/transport/handler.go

```diff
@@ -134,19 +134,41 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		_ = r.Body.Close()
 	}()
 
-	// Buffer the body for later use to track slow queries.
-	var buf bytes.Buffer
-	r.Body = http.MaxBytesReader(w, r.Body, f.cfg.MaxBodySize)
-	r.Body = io.NopCloser(io.TeeReader(r.Body, &buf))
+	// Store the body contents in a seeker, so we can read it multiple times.
+	seeker, err := readIntoReadCloseSeeker(http.MaxBytesReader(w, r.Body, f.cfg.MaxBodySize))
+	if err != nil {
+		writeError(w, err)
+		return
+	}
+	r.Body = seeker
+
+	// Parse the form, as it's needed to build the activity for the activity-tracker.
+	if err := r.ParseForm(); err != nil {
+		writeError(w, apierror.New(apierror.TypeBadData, err.Error()))
+		return
+	}
+
+	// Store the r.Form contents copy, as middlewares may modify it (like setting the default values).
+	params := copyValues(r.Form)
+
+	// Seek the body back to the beginning, so it can be read again if it's used to forward the request through a roundtripper.
+	if _, err := seeker.Seek(0, io.SeekStart); err != nil {
```

Comment:
Seeker allows next user of `r.Body` to just seek it back instead of having to read it again to a new buffer in order to replace it.

Reply:
I don't think there is another user of r.Body doing that (except http -> httpgrpc conversion, which reads body into new buffer again).

Reply:
It would also be very strange if there was code checking if `r.Body` implements Seeker. That's very unusual interface for HTTP request body to implement, given that it's typically streamed over network.

Reply:
I think once again we make things little bit more complicated and generic than they need to be. But my intention is not to block the PR on this, feel free to merge it as-is.